test/ServerInvoiceTest.php

```diff
@@ -280,4 +280,77 @@ public function testCustomFields()
 //        $cfs = $invoice->getCustomFields($this->tenant->getTenantHeaders());
 //        $this->assertEquals(2, count($cfs));
     }
+
+
+    /**
+     * @test
+     */
+    public function itShouldCreateInvoiceWhenSubscribedOnBasePlan()
+    {
+        $subscriptionBase = new Subscription();
+        $subscriptionBase->setAccountId($this->account->getAccountId());
+        $subscriptionBase->setPlanName('sports-monthly');
+        $subscriptionBase = $this->client->getSubscriptionApi()->createSubscription($subscriptionBase, self::USER, self::REASON, self::COMMENT);
+        self::assertSame($this->account->getAccountId(), $subscriptionBase->getAccountId());
+        self::assertSame('sports-monthly', $subscriptionBase->getPlanName());
+
+        $invoices = $this->client->getAccountApi()->getInvoicesForAccount(
+            $this->account->getAccountId(),
+            null,
+            $withItems = 'true'
+        );
+        $this->assertCount(1, $invoices);
+        $this->assertSame(0.0, $invoices[0]->getAmount());
+
+        $this->clock->addDays(35);
+
+        $invoicesNextMonth = $this->client->getAccountApi()->getInvoicesForAccount(
+            $this->account->getAccountId(),
+            null,
+            $withItems = 'true'
+        );
+        $this->assertCount(2, $invoicesNextMonth);
+        $this->assertSame(0.0, $invoicesNextMonth[0]->getAmount());
+        $this->assertSame(500.0, $invoicesNextMonth[1]->getAmount());
+    }
+
+    /**
+     * @test
+     */
+    public function itShouldCreateInvoiceWhenSubscribedOnBasePlanAndConsumableAddon()
+    {
+        $subscriptionBase = new Subscription();
+        $subscriptionBase->setAccountId($this->account->getAccountId());
+        $subscriptionBase->setPlanName('sports-monthly');
+        $subscriptionBase = $this->client->getSubscriptionApi()->createSubscription($subscriptionBase, self::USER, self::REASON, self::COMMENT);
+        self::assertSame($this->account->getAccountId(), $subscriptionBase->getAccountId());
+        self::assertSame('sports-monthly', $subscriptionBase->getPlanName());
+
+        $subscriptionGas = new Subscription();
+        $subscriptionGas->setAccountId($this->account->getAccountId());
+        $subscriptionGas->setPlanName('gas-monthly');
+        $subscriptionGas->setBundleId($subscriptionBase->getBundleId());
+        $subscriptionGas = $this->client->getSubscriptionApi()->createSubscription($subscriptionGas, self::USER, self::REASON, self::COMMENT);
+        self::assertSame($this->account->getAccountId(), $subscriptionGas->getAccountId());
+        self::assertSame('gas-monthly', $subscriptionGas->getPlanName());
+
+        $invoices = $this->client->getAccountApi()->getInvoicesForAccount(
+            $this->account->getAccountId(),
+            null,
+            $withItems = 'true'
+        );
+        $this->assertCount(1, $invoices);
+        $this->assertSame(0.0, $invoices[0]->getAmount());
+
+        $this->clock->addDays(35);
```

Comment:
After moving the clock, asynchronous processing of subscriptions and invoice generation will happen.

We have some logic to ensure all asynchronous jobs are fully processed before returning (provided the [timeout](https://github.com/killbill/killbill/blob/6457b485fb32a182c76197a83f428123331f1380/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/TestResource.java#L198) is long enough), but the heuristic isn't always enough depending on the catalog, config, etc.

A good practice in these tests is to wait on the client side for an expected condition (e.g. a new invoice being generated) and fail the test if this doesn't happen after a timeout. See https://github.com/killbill/killbill-integration-tests/blob/2eec63809331bb0cb8217554542740352d5666aa/killbill-integration-tests/core/test_entitlement_create.rb#L29 for an example in our Ruby tests.